api/document/json_importer/annotations.py

```diff
@@ -36,29 +36,51 @@ def footnote_citation(cursor: JSONAwareCursor, content: JsonDict,
 @annotator
 def external_link(cursor: JSONAwareCursor, content: JsonDict,
                   start: int) -> ExternalLink:
+    text = get_content_text(content['inlines'])
     return ExternalLink(
         doc_node=cursor.model, start=start,
-        end=start + len(content['text']), href=content['href']
+        end=start + len(text), href=content['href']
     )
 
 
 AnnotationDict = DefaultDict[Type[Annotation], List[Annotation]]
 
 
+def find_annotations(items: List[JsonDict],
+                     start: int=0) -> Iterator[Tuple[JsonDict, int]]:
+    for content in items:
+        if content['content_type'] == '__text__':
+            start += len(content['text'])
+        else:
+            yield (content, start)
+            for icontent, istart in find_annotations(content['inlines'],
+                                                     start):
+                yield (icontent, istart)
+                start = istart
```

Comment:
Do you think we could make this and `derive_annotations` simpler by combining them somehow? This function seems very abstract; I speculate that the logic would be easier to follow if we were yielding annotations instead of `Tuple(JsonDict, int)` pairs.